walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,83 @@
 
 package com.google.sps;
 
+import java.util.ArrayList;
 import java.util.Collection;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    List<TimeRange> requiredTimeRanges = findTimeSlotsForNewMeeting(
+        filterRelevantEvents(events, request.getAttendees()),
+        request.getDuration());
+    List<TimeRange> optionalTimeRanges = findTimeSlotsForNewMeeting(
+        filterRelevantEvents(events, request.getOptionalAttendees()),
+        request.getDuration());
+    List<TimeRange> intersection = findTimeRangesIntersections(
+        requiredTimeRanges, optionalTimeRanges, request.getDuration());
+    if (intersection.isEmpty()) {
+      return requiredTimeRanges;
+    }
+    return intersection;
+  }
+
+  private static List<TimeRange> findTimeRangesIntersections(List<TimeRange> required,
+                                                             List<TimeRange> optional,
+                                                             long duration) {
+    List<TimeRange> intersection = new ArrayList<>();
+    for (TimeRange requiredTimeRange : required) {
+      for (TimeRange optionalTimeRange : optional) {
+        createIntersection(requiredTimeRange, optionalTimeRange)
+            .filter(timeRange -> timeRange.duration() >= duration)
+            .ifPresent(intersection::add);
+      }
+    }
+    return intersection;
```

Comment:
quite brute forced, is there another possiblity?

Reply:
There is another solution using two pointers algorithm, which improves complexity from O(n^2) to O(n), but in this case I'm not sure if this is a better way. Its implementation will be less straightforward and since number of TimeRange objects won't exceed 1140 (number of minutes in a day), the difference in time efficiency won't be high.
What do you think about it? Is this trade-off worth a try?

The complexity bottleneck of this whole solution lies in somewhere else -- filtering events in which there is at least one attendee from query meeting. Since the query method is public, one can provide really big list of events, containing lots of attendees. The method goes one by one checking every event, in each event it searches for at least one required attendee. Some optimisations can be made -- for example stopping checks after finding first required attendee. But in case someone provides malicious data it will be even slower than current solution.

Reply:
Great explanation, sounds good :+1: